controllers/operator/apigateway_controller.go

```diff
@@ -211,7 +211,7 @@ func (r *APIGatewayReconciler) reconcileFinalizer(ctx context.Context, apiGatewa
 
 	if apiGatewayCR.IsInDeletion() && hasFinalizer(apiGatewayCR) {
 		apiRulesFound, err := apiRulesExist(ctx, r.Client)
-		if err != nil {
+		if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
We should have a test to confirm that this implementation actually fixes a bug. Furthermore, I think this is already covered by the test [Should delete API-Gateway CR if there are no blocking resources](https://github.com/kyma-project/api-gateway/blob/23271ca5c8318a8c4993c1de39db0d19161a224b/controllers/operator/apigateway_controller_test.go#L232) and the test is not failing.
The [docs about the List API](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client@v0.17.2#Reader) are not 100% clear, but it looks like no error is returned when nothing is found, what makes sense for a REST API.

Reply:
The problem is what happens when the CR you refer to doesn't exist. The issue occurs only when the deletion is in progress. The error message clearly states `could not retrieve the complete list` which indicates the problem occurs when you try to list the Resources when the Definition doesn't exist. That's the only place where this condition occurs. The rest is in the tests.

I already looked at the current implementation of tests and you are correct. However I couldn't find a solution how to make that in unit test. You have to ensure that ORY CRDs are not registered. Not to mention I still can't hook the debugger to those tricky ginkgo tests, so I cannot ensure the test covers this specific case.